### ISSUE TITLE
Bugs/Improvements

### ISSUE THREAD

**Issue by hamDrambus** [hamDrambus/ElectronDriftThreeDim#1]
//L - low priority, M - medium, H - high.
//+ --'to add'
1)M+ settings store relative paths? Current situation is very error-prone
2)L+ each test to settings as optional module
3)L remove FactorHelper and other unused functions? Maybe leave Tforward/backward and Pback.
5)L Cmake (3 dependencies: boost, CERN ROOT and gnuplot. The latter one is optional and is for displaying results)
6)L+ comprehensive and extensive readme. With diagram.
8)L paths implementation with boost.
9)L multithread table construction.
11)L GPU acceleration
12)M+ Diffusion coefficients
14)H Optimization:
&nbsp;&nbsp;&nbsp;&nbsp;4) Remove Event argument from functions of Manager
&nbsp;&nbsp;&nbsp;&nbsp;5) Decrease energy and angle discretization
&nbsp;&nbsp;&nbsp;&nbsp;6) Unify argon spin_flip and spin_nonflip amplitude functions.
&nbsp;&nbsp;&nbsp;&nbsp;7) Inline setters/getters all across the program (InelasticProcess, ...)  
&nbsp;&nbsp;&nbsp;&nbsp;9) Mixture::GetXSIntegral: search 2 values in the table simultaneously for the same Ey. (In one loop until possible, when branches diverge, search separately. find_Ey_index is called only once).  
&nbsp;&nbsp;&nbsp;&nbsp;10) + bucketing to all search algorithms.
&nbsp;&nbsp;&nbsp;&nbsp;12) + caching in FunctionTable and DataVector (in find E/Ey and operator()). It will also prevent calling TotalXS(E) for argon and argon2 twice because they share the same XS tables.
17)L change INVOKE_GNUPLOT for windows (use CreateProcess)
18)M implement DataVector using templates. 2Dim table can then be implemented as DataVector of DataVectors. get_Y_indices won't work in general case though.
19)L + every thread prints their progress in single line (updating it).
20)H replace std::pow in PolynimialFit to custom pown(double, uint) (from same place in Data_processing)
21)H Data leak when doing multiple runs. (previous ttrees are not cleared after writing to file)
22)M In order to increase data capacity, cross-section and other data table should be unloaded before merging the ttrees

DONE:
~~0)L refactor DataVector and PDF_routine. Replace repeating code by PDF_routine~~
~~0.1)L optimize PDF_routine (construct cdf concurrently with push())~~
~~4)L Refactor argon_cross.cpp(h) - filename, too much code in one place and quite complicated.~~
~~7)H+ drift in mixture. + e+Ar2(1%) -> Ar- (long life time) + Ar mechanism of slow component.~~
~~10)H profiler for ~(eclipse)~ ubuntu~~
~~13)H Make ArgonVanDerWaals inherit from argon particle, and use double XS for all processes + dissociation attachment.~~
14)H Optimization:
&nbsp;&nbsp;&nbsp;&nbsp;[optz1]~~0) Add Particle->GetTotalXS using tables. (mb not even saving to file). Selecting scattering particle uses too much calculations of individual XS.~~
&nbsp;&nbsp;&nbsp;&nbsp;[not relevant]~~1) boost::thread_specific_ptr<std::deque<...>> in Particles in order to avoid multiple allocations of cross_section vector to calculate process~~
&nbsp;&nbsp;&nbsp;&nbsp;[optz1]~~2) Do not calculate cross sections for all processes at once. Store total cross-section and first calculate only XS[0], then check R<XS[0]/XS_T, then calculate XS[1] and so on. Because XS[0] is dominant process and calculating XS is quite expensive, this will boost the performance. The issue is that XS_T which will be taken from table will not be the same as sum of XS[i] precisely.~~
&nbsp;&nbsp;&nbsp;&nbsp;[not relevant]~~3) Try change Es[Ey_index] to vector<double>* = &Es[Ey_index] in FunctionTable::find_E_indexes in order to avoid repeated indexation (maybe compiler does it though)~~
&nbsp;&nbsp;&nbsp;&nbsp;[optz2 - reworked]~~8) Find_E_indexes_by_value: move E_index = (left+right)/2 inside~~
&nbsp;&nbsp;&nbsp;&nbsp;[otpz2]~~11) Optimize binary search algorithm (change if conditions, minimize calls).~~
~~15)L+ #define USE_ROOT_ to remove all root dependencies for profiling on windows (can't build project using ROOT6 there). Maybe try with ROOT5.~~
[optz2]~~16)L Rename functions in 2D tables (use X-Y instead of obscure E-Ey)~~
